### DISCUSSION TITLE
Compile with `-O2` instead of `-O3`

### DISCUSSION THREAD

**Discussion by brandtbucher** [faster-cpython/ideas#140] [Ideas]
On a hunch, I tried compiling CPython with `-O2`… which results in a 1% speedup.

If I had to attribute this to anything, I’d guess that it’s because `-O3`’s aggressive inlining and loop unrolling hurts i-cache locality (the `-O2` executable is ~3.5% smaller than the `-O3` one).

<details>

| Benchmark               | main    | O2                    |
|-------------------------|:-------:|:---------------------:|
| pickle_dict             | 29.5 us | 27.0 us: 1.09x faster |
| regex_v8                | 25.5 ms | 23.6 ms: 1.08x faster |
| regex_dna               | 223 ms  | 207 ms: 1.08x faster  |
| regex_effbot            | 3.31 ms | 3.08 ms: 1.07x faster |
| unpack_sequence         | 46.0 ns | 43.9 ns: 1.05x faster |
| json_loads              | 28.0 us | 27.1 us: 1.03x faster |
| chaos                   | 83.8 ms | 81.3 ms: 1.03x faster |
| telco                   | 6.42 ms | 6.24 ms: 1.03x faster |
| float                   | 86.9 ms | 85.3 ms: 1.02x faster |
| fannkuch                | 425 ms  | 418 ms: 1.02x faster  |
| json_dumps              | 12.8 ms | 12.6 ms: 1.01x faster |
| pickle_list             | 4.64 us | 4.57 us: 1.01x faster |
| crypto_pyaes            | 90.8 ms | 89.6 ms: 1.01x faster |
| unpickle                | 13.8 us | 13.6 us: 1.01x faster |
| mako                    | 12.7 ms | 12.5 ms: 1.01x faster |
| spectral_norm           | 119 ms  | 117 ms: 1.01x faster  |
| meteor_contest          | 107 ms  | 106 ms: 1.01x faster  |
| python_startup          | 9.47 ms | 9.39 ms: 1.01x faster |
| python_startup_no_site  | 6.36 ms | 6.31 ms: 1.01x faster |
| scimark_fft             | 358 ms  | 356 ms: 1.01x faster  |
| xml_etree_parse         | 156 ms  | 155 ms: 1.01x faster  |
| unpickle_pure_python    | 268 us  | 266 us: 1.01x faster  |
| hexiom                  | 7.78 ms | 7.75 ms: 1.00x faster |
| pickle                  | 10.2 us | 10.1 us: 1.00x faster |
| pidigits                | 188 ms  | 188 ms: 1.00x slower  |
| chameleon               | 7.56 ms | 7.61 ms: 1.01x slower |
| pathlib                 | 19.2 ms | 19.3 ms: 1.01x slower |
| logging_simple          | 6.72 us | 6.76 us: 1.01x slower |
| scimark_monte_carlo     | 80.9 ms | 81.5 ms: 1.01x slower |
| raytrace                | 350 ms  | 353 ms: 1.01x slower  |
| sympy_expand            | 512 ms  | 517 ms: 1.01x slower  |
| xml_etree_process       | 61.2 ms | 61.9 ms: 1.01x slower |
| xml_etree_iterparse     | 107 ms  | 108 ms: 1.01x slower  |
| logging_format          | 7.32 us | 7.40 us: 1.01x slower |
| django_template         | 39.4 ms | 39.9 ms: 1.01x slower |
| richards                | 59.5 ms | 60.4 ms: 1.02x slower |
| logging_silent          | 124 ns  | 126 ns: 1.02x slower  |
| go                      | 175 ms  | 178 ms: 1.02x slower  |
| pyflate                 | 541 ms  | 550 ms: 1.02x slower  |
| scimark_sparse_mat_mult | 4.95 ms | 5.09 ms: 1.03x slower |
| unpickle_list           | 4.78 us | 4.92 us: 1.03x slower |
| Geometric mean          | (ref)   | 1.01x faster          |
</details>

**Comment by tiran** [comment]
Red Hat's and Fedora's Python are compiled with ``-O2``. Florian Weimar's blog post https://developers.redhat.com/blog/2018/03/21/compiler-and-linker-flags-gcc also recommends ``-O2`` for most applications:

> For many applications, -O2 is a good choice because the additional inlining and loop unrolling introduced by -O3 increases the instruction cache footprint, which ends up reducing performance

Florian also suggests to try ``-ftree-loop-vectorize`` with ``-O2``:

> For some applications -O3 or -O2 -free-loop-vectorize will provide a significant speed boost. By default, GCC does not perform loop vectorization.

(The blog post has a typo. I have informed Florian that the correct spelling is ``-ftree-loop-vectorize``.

**Comment by brandtbucher** [comment]
Awesome, thanks for the link! I'll find time today to try that out.

**Comment by brandtbucher** [comment]
Looks like `-O2 -ftree-loop-vectorize` is ~1% slower than vanilla `-O3`:

<details>

```
Slower (25):
- unpack_sequence: 43.7 ns +- 0.9 ns -> 48.8 ns +- 2.9 ns: 1.12x slower
- pickle_dict: 26.8 us +- 0.1 us -> 28.1 us +- 0.1 us: 1.05x slower
- regex_effbot: 3.15 ms +- 0.03 ms -> 3.25 ms +- 0.04 ms: 1.03x slower
- pickle_list: 4.48 us +- 0.04 us -> 4.63 us +- 0.04 us: 1.03x slower
- spectral_norm: 107 ms +- 1 ms -> 111 ms +- 2 ms: 1.03x slower
- scimark_monte_carlo: 77.5 ms +- 0.8 ms -> 79.9 ms +- 1.3 ms: 1.03x slower
- regex_dna: 212 ms +- 1 ms -> 217 ms +- 1 ms: 1.02x slower
- scimark_sor: 154 ms +- 2 ms -> 158 ms +- 2 ms: 1.02x slower
- richards: 56.9 ms +- 0.9 ms -> 57.9 ms +- 1.5 ms: 1.02x slower
- deltablue: 4.85 ms +- 0.06 ms -> 4.93 ms +- 0.06 ms: 1.02x slower
- go: 165 ms +- 1 ms -> 168 ms +- 2 ms: 1.02x slower
- logging_silent: 116 ns +- 3 ns -> 117 ns +- 1 ns: 1.01x slower
- float: 80.5 ms +- 0.9 ms -> 81.4 ms +- 0.8 ms: 1.01x slower
- regex_v8: 23.2 ms +- 0.1 ms -> 23.4 ms +- 0.1 ms: 1.01x slower
- sqlite_synth: 2.74 us +- 0.04 us -> 2.76 us +- 0.07 us: 1.01x slower
- mako: 11.9 ms +- 0.1 ms -> 12.0 ms +- 0.1 ms: 1.01x slower
- pickle_pure_python: 367 us +- 3 us -> 371 us +- 3 us: 1.01x slower
- scimark_sparse_mat_mult: 4.84 ms +- 0.08 ms -> 4.89 ms +- 0.10 ms: 1.01x slower
- tornado_http: 109 ms +- 3 ms -> 110 ms +- 3 ms: 1.01x slower
- pathlib: 19.0 ms +- 0.3 ms -> 19.1 ms +- 0.2 ms: 1.01x slower
- 2to3: 271 ms +- 1 ms -> 272 ms +- 1 ms: 1.00x slower
- regex_compile: 144 ms +- 2 ms -> 145 ms +- 1 ms: 1.00x slower
- dulwich_log: 67.6 ms +- 0.3 ms -> 67.8 ms +- 0.3 ms: 1.00x slower
- python_startup: 8.51 ms +- 0.01 ms -> 8.52 ms +- 0.01 ms: 1.00x slower
- python_startup_no_site: 5.90 ms +- 0.00 ms -> 5.90 ms +- 0.00 ms: 1.00x slower

Faster (15):
- xml_etree_iterparse: 107 ms +- 3 ms -> 105 ms +- 1 ms: 1.02x faster
- nqueens: 90.1 ms +- 0.7 ms -> 88.2 ms +- 0.9 ms: 1.02x faster
- fannkuch: 427 ms +- 4 ms -> 418 ms +- 4 ms: 1.02x faster
- json_dumps: 12.6 ms +- 0.1 ms -> 12.4 ms +- 0.1 ms: 1.02x faster
- unpickle_list: 5.05 us +- 0.04 us -> 4.95 us +- 0.06 us: 1.02x faster
- pidigits: 190 ms +- 0 ms -> 188 ms +- 0 ms: 1.01x faster
- crypto_pyaes: 87.5 ms +- 0.6 ms -> 86.6 ms +- 0.6 ms: 1.01x faster
- hexiom: 7.53 ms +- 0.12 ms -> 7.46 ms +- 0.07 ms: 1.01x faster
- sqlalchemy_imperative: 18.6 ms +- 0.5 ms -> 18.4 ms +- 0.2 ms: 1.01x faster
- xml_etree_parse: 155 ms +- 3 ms -> 154 ms +- 1 ms: 1.01x faster
- django_template: 37.2 ms +- 0.5 ms -> 37.0 ms +- 0.5 ms: 1.01x faster
- pickle: 9.97 us +- 0.16 us -> 9.89 us +- 0.06 us: 1.01x faster
- xml_etree_generate: 80.4 ms +- 0.7 ms -> 80.0 ms +- 0.9 ms: 1.01x faster
- logging_simple: 6.15 us +- 0.08 us -> 6.12 us +- 0.10 us: 1.01x faster
- raytrace: 331 ms +- 2 ms -> 330 ms +- 2 ms: 1.00x faster

Benchmark hidden because not significant (18): chameleon, chaos, json_loads, logging_format, meteor_contest, nbody, pyflate, scimark_fft, scimark_lu, sqlalchemy_declarative, sympy_expand, sympy_integrate, sympy_sum, sympy_str, telco, unpickle, unpickle_pure_python, xml_etree_process

Geometric mean: 1.01x slower
```

</details>

**Comment by tiran** [comment]
So ``-O2`` > ``-O3`` > ``-O2 -ftree-loop-vectorize``?
